astropy/io/votable/tree.py

```diff
@@ -4378,7 +4381,11 @@ def to_xml(
                     self.resources,
                 ]
                 if kwargs["version_1_2_or_later"]:
-                    element_sets[0] = self.groups
+                    element_sets.append(self.groups)
+                    if not kwargs["version_1_4_or_later"]:
+                        # COOSYS was deprecated in v1.2, and undeprecated in v1.4
+                        del element_sets[0]
```

Comment:
I see that this deletion of the `coordinate_systems` was just preserving the previous behavior for the v1.2 and v1.3 while allowing it to go through for >=v1.4.  Then the new deletion above for `RESOURCE` is intending to match this behavior.

This is a pretty quirky case!  After more thinking about this, I believe that the original replacement here of coordinate_systems with groups was a mistake.  In fact `COOSYS` never became illegal, it was just deprecated, so it was ok to keep quietly supporting it.  

It's interesting that the original deletion here for `VOTABLE` was not mirrored for `RESOURCE`, but it definitely left us in an inconsistent state, and with the bug you caught in #17348.  I think the better fix here would be to just do the `append()` of the groups, but not the deletion, then not do the deletion in `RESOURCE`.

I apologize that I didn't write this before you went to the trouble of writing the extra version-specific testing!

Reply:
Not a problem, the updated tests were pretty straightforward.

I've gone ahead and implemented the change to keep `COOSYS` regardless of version and updated the truth reference to match the expected behavior. I rebased the whole PR to remove the unnecessary commits.